### PR TITLE
include/interval_set: parameterize by map type and kill btree_interval_set.h

### PR DESCRIPTION
**Pull request by liewegas**
The goal here is to be able to use boost::flat_map<> for something new (new_{removed,purged}_snaps in OSDMap).

### CODE REVIEW

File: src/include/interval_set.h [hunk 1]
```diff
@@ -225,8 +225,8 @@ class interval_set {
     return p;
   }
   
-  typename std::map<T,T>::iterator find_adj_m(T start) {
-    typename std::map<T,T>::iterator p = m.lower_bound(start);
+  typename Map::iterator find_adj_m(T start) {
+    typename Map::iterator p = m.lower_bound(start);
```
Comment (adamemerson) [inline src/include/interval_set.h]: At this point you may as well use auto.

File: src/include/btree_interval_set.h [hunk 2]
```diff
@@ -1,625 +0,0 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (adamemerson) [inline src/include/btree_interval_set.h]: Yay.

File: src/test/common/test_interval_set.cc [hunk 3]
```diff
@@ -32,7 +33,9 @@ class IntervalSetTest : public ::testing::Test {
 typedef ::testing::Types<
   interval_set<IntervalValueType>,
   interval_set<IntervalValueType,
-	       btree::btree_map<IntervalValueType,IntervalValueType>>
+	       btree::btree_map<IntervalValueType,IntervalValueType>>,
+  interval_set<IntervalValueType,
+	       boost::container::flat_map<IntervalValueType,IntervalValueType>>
   > IntervalSetTypes;
```
Comment (adamemerson) [inline src/test/common/test_interval_set.cc]: Excellent.

**Comment by branch-predictor** [comment]
Can't do proper review/testing at this time, but I'm more than happy to see flat_map being actually used in right place. Wouldn't mind removing std::map from interval_set entirely.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-10-30_02:29:13-rados-wip-sage-testing-2017-10-29-1619-distro-basic-smithi/

**Pull request merged**